cpp/integ/date_time_test.cc

```diff
@@ -140,4 +141,190 @@ TEST_F(TestProjector, TestTime) {
   EXPECT_ARROW_ARRAY_EQUALS(exp_dd_from_ts, outputs.at(3));
 }
 
-}  // namespace gandiva
+TEST_F(TestProjector, TestTime) {
+  auto field0 = field("f0", time32(arrow::TimeUnit::MILLI));
+  auto schema = arrow::schema({field0});
+
+  auto field_min = field("mm", int64());
+  auto field_hour = field("hh", int64());
+
+  // extract day and hour from time32
+  auto time2min_expr = TreeExprBuilder::MakeExpression(
+    "extractMinute",
+    {field0},
+    field_min);
+  auto time2hour_expr = TreeExprBuilder::MakeExpression(
+    "extractHour",
+    {field0},
+    field_hour);
+
+  std::shared_ptr<Projector> projector;
+  Status status = Projector::Make(
+    schema,
+    {time2min_expr, time2hour_expr},
+    pool_,
+    &projector);
+  ASSERT_TRUE(status.ok());
+
+  // create input data
+  int num_records = 4;
+  auto validity = { true, true, true, true };
+  std::vector<int32_t> field_data = {
+    MillisInDay(5, 35, 25, 0), // 5:35:25
+    MillisInDay(0, 59, 0, 0),  // 0:59:12
+    MillisInDay(12, 30, 0, 0), // 12:30:0
+    MillisInDay(23, 0, 0, 0)   // 23:0:0
+  };
+  auto array = MakeArrowTypeArray<arrow::Time32Type, int32_t>(
+    time32(arrow::TimeUnit::MILLI),
+    field_data,
+    validity);
+
+  // expected output
+  auto exp_min = MakeArrowArrayInt64({35, 59, 30, 0}, validity);
+  auto exp_hour = MakeArrowArrayInt64({5, 0, 12, 23}, validity);
+
+  // prepare input record batch
+  auto in_batch = arrow::RecordBatch::Make(schema, num_records, {array});
+
+  // Evaluate expression
+  arrow::ArrayVector outputs;
+  status = projector->Evaluate(*in_batch, &outputs);
+  EXPECT_TRUE(status.ok());
+
+  // Validate results
+  EXPECT_ARROW_ARRAY_EQUALS(exp_min, outputs.at(0));
+  EXPECT_ARROW_ARRAY_EQUALS(exp_hour, outputs.at(1));
+}
+
+TEST_F(TestProjector, TestTimestampDiff) {
```

Comment:
don't need. to change now but going forward, we should add gandiva tests mainly for new data-types, new literals and new control expressions. The fn specific tests should go into the precompiled unit tests as far as possible (eg. precompiled/time_test.cc) since they are easier to write/debug.